Packages/com.unity.inputsystem/InputSystem/Plugins/Visualizers/InputControlVisualizer.cs

```diff
@@ -0,0 +1,361 @@
+            {
+                var component = s_EnabledInstances[i];
+                if (component.m_Control.device != device || component.m_Visualizer == null)
+                    continue;
+
+                component.OnEventImpl(eventPtr);
+            }
+        }
+
+        private unsafe void OnEventImpl(InputEventPtr eventPtr)
+        {
+            switch (m_Visualization)
+            {
+                case Mode.Value:
+                {
+                    var statePtr = m_Control.GetStatePtrFromStateEvent(eventPtr);
+                    if (statePtr == null)
+                        return; // No value for control in event.
+                    var value = m_Control.ReadValueFromStateAsObject(statePtr);
+                    m_Visualizer.AddSample(value, eventPtr.time);
+                    break;
+                }
+
+                case Mode.Events:
+                {
+                    var visualizer = (VisualizationHelpers.TimelineVisualizer)m_Visualizer;
+                    var frame = (int)InputUpdate.s_UpdateStepCount;
+                    ref var valueRef = ref visualizer.GetOrCreateSample(0, frame);
+                    var value = valueRef.ToInt32() + 1;
+                    valueRef = value;
+                    visualizer.limitsY =
+                        new Vector2(0, Mathf.Max(value, visualizer.limitsY.y));
+                    break;
+                }
+
+                case Mode.MaximumLag:
+                {
+                    var visualizer = (VisualizationHelpers.TimelineVisualizer)m_Visualizer;
+                    var lag = (Time.realtimeSinceStartup - eventPtr.time) * 1000; // In milliseconds.
+                    var frame = (int)InputUpdate.s_UpdateStepCount;
+                    ref var valueRef = ref visualizer.GetOrCreateSample(0, frame);
+
+                    if (lag > valueRef.ToDouble())
+                    {
+                        valueRef = lag;
+                        if (lag > visualizer.limitsY.y)
+                            visualizer.limitsY = new Vector2(0, Mathf.Ceil((float)lag));
+                    }
+                    break;
+                }
+
+                case Mode.Bytes:
+                {
+                    var visualizer = (VisualizationHelpers.TimelineVisualizer)m_Visualizer;
+                    var frame = (int)InputUpdate.s_UpdateStepCount;
+                    ref var valueRef = ref visualizer.GetOrCreateSample(0, frame);
+                    var value = valueRef.ToInt32() + eventPtr.sizeInBytes;
+                    valueRef = value;
+                    visualizer.limitsY =
+                        new Vector2(0, Mathf.Max(value, visualizer.limitsY.y));
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Determines which aspect of the control should be visualized.
+        /// </summary>
+        public enum Mode
+        {
+            None = 0,
```

Comment:
where do these enum integer values come from?

Reply:
I was adding and removing entries as I was writing this and after I had a bunch of these components saved in scenes, was too lazy to go and update them after each change so ended up hardcoding the enum values :)

Reply:
So I guess it would make sense to remove them now?

Reply:
Does it really matter? Sure I can remove them and update the prefabs I created. Is it important?